### DISCUSSION TITLE
Add useQuery

### DISCUSSION THREAD

**Discussion by muezz** [next-safe-action/next-safe-action#191] [Ideas]
In my last project, I was using tRPC and react-query to handle all of my mutations and query logic. In my current project, I decided to avoid the extremely complicated and bloated set up of my last project. 

I have been using this library for a few days and everything was fine as long as I was using the next safe actions to fetch data in server components or mutating data in my client components. Now, I am in a scenario where I want to fetch something in a client component while retaining similar functionality as react-query.

I think an equivalent of `useQuery` would be amazing with config options like `enabled`, `isLoading`, `data`, `error` and `reset`.

**Comment by TheEdoRan** [comment]
Is [`useAction`](https://next-safe-action.dev/docs/execution/hooks/useaction) not good enough for your use case?

**Reply by fitimbytyqi** [reply]
@muezz next-safe-actions is meant to be used for mutations. I wouldn't suggest using actions for querying because of server action nuances https://tigerabrodi.blog/nuances-of-server-actions-in-nextjs.

What I've started doing in my BFF next apps is (trpc & react query for fetching) & next-safe-actions for mutating. 

Seperation of concerns.

**Reply by muezz** [reply]
@TheEdoRan `useAction` is only for mutating the data. I mean it does not trigger on first render like `useQuery` from react-query.

For example opening a page that shows the products and the fetching only starts after the user is on the page. Therefore, they will see a loading state. I understand that in most cases you can just fetch the products in the parent server component but I am in a scenario where I want the ability to do it on the client.

I hope that makes sense.

**Reply by muezz** [reply]
@fitimbytyqi I don't fully understand the issue here though. If I have a normal "server action" (without binding it to a next safe client), I can safely use it to fetch data in a server component, right? Why should it not be possible in client components? I mean tRPC does it. As far as I know, their queries and mutations behave like server components as in they are API routes that live in the server and are invoked from the client.

The reason I don't want to user tRPC with react query is the unnecessary amount of setup thats required for it. And to be fair, if I did do that setup, what is the point of using next-safe-actions if I can just use mutations from trpc+react-query?

My knowledge is limited so if I misunderstand a concept, please feel free to correct me.

**Reply by fitimbytyqi** [reply]
@muezz Sure you could just pass a server action into useQuery's queryFn, that works but it has limitations.

If you have 2 useQuerys that trigger on component mount, the latter will always wait for the first one to finish because server actions dont run in parallel.

So in this case if the first server action needs 4s / second server action needs 4s as well, you will see the data from the second query after 8 seconds.

Server actions create a POST request under the hood. But ideally we need a GET request right? In this case we have to write GET route handlers, but thats terrible DX if we are using Nextjs as BFF, thats where trpc comes in, It lets you create functions which generate route handlers. 

Ofc I would love to have one for all and Im not liking the setup hell of trpc, react query and now next-safe-actions but it gets the job done.

**Reply by muezz** [reply]
@fitimbytyqi Hmmm. That is interesting. I think that is not the case. At least in react-wuery, if you have two `useQuery` hooks in a component, they are not blocked by each other. I mean the second one does not wait for the first one to finish (assuming they it does not rely on the result from the first).

If you mean in a server component putting two server actions like so:
```tsx
export default async function TestPage() {
  const tokens = await getCreditsCount();
  const products = await getProducts();

  return (
    <div>
      -- my custom content --
    </div>
  );
}
```
then yes, the products will only be fetched after you get the credits count. This can be "fixed" by using `Promise.all()`.

All this is besides the point. This is in a server component. I am requesting the functionality of `useQuery` from react-query from this library so I am not forced to do the hellish set up of tRPC+react query.

**Reply by fitimbytyqi** [reply]
@muezz I wouldn't suggest asking the author to recreate the wheel xD, there's already react query and its a beast.

Also, here's the sandbox I made for you to showcase the server action limitations.

You need to wait 8 seconds for both to show up.

https://codesandbox.io/p/devbox/exciting-pateu-s36h4w?file=%2Fapp%2F_components%2Fclient-comp.tsx%3A7%2C44

**Reply by muezz** [reply]
@fitimbytyqi I understand if it is too much work for a solo maintainer to do it for free. It was only a request. I would be open to doing it myself if the maintainer is open to the idea.

I honestly did not expect that to take 8 seconds. See this [link](https://tanstack.com/query/latest/docs/framework/react/guides/parallel-queries) from the react-query docs. I guess I was wrong but I am truly surprised.

**Reply by Kiranism** [reply]
The alt package Zsa has a wrapper around React Query https://zsa.vercel.app/docs/react-query

**Reply by muezz** [reply]
@Kiranism That looks really promising. This is precisely the kind of thing I needed. Thanks. I will look into it.

**Reply by TheEdoRan** [reply]
@muezz it seems that zsa solved your problem, which is great. I personally think that adding a `useQuery` hook is out of the scope of this library, since Server Actions should be used for mutations and only work via POST method. That said, I could maybe work on a TanStack Query integration in the future, if users want that.

**Reply by markomitranic** [reply]
I came here trying to figure out why the author made the decision to not use React/Tanstack Query for this, and instead seems to have implemented their own query-inspired solution :D weird. Would love to learn more about why if possible. The only assumption I could make was that they wanted something built-in as a baby step, so that most users have the batteries they need out of the box, while keeping it open for manual integration with a real library. The docs wording gives me that vibe.

To answer the OP - Server actions by default work entirely fine with Tanstack Query's `useMutation`, so I'd expect `next-safe-action` actions to work exactly the same with `useMutation`.

```ts
"use client";
import { useMutation } from "@tanstack/react-query";

export function ClientComponent() {
    const { mutate, isPending, data } = useMutation({ mutationFn: myServerActionFn });
```

Has all the bells and whistles. Is there any reason you couldn't do this?

**Reply by muezz** [reply]
@TheEdoRan I often use server actions to fetch data in server components. They are especially nice when I want to make the same call in different routes and dont want to expose an API route. I feel like this should be possible to do from the client side as well.

In any case, I am gonna stick with this library for now. Hopefully this gets added if more people mention a need for it.

**Reply by prpanto** [reply]
> The alt package Zsa has a wrapper around React Query https://zsa.vercel.app/docs/react-query

Thanks very much. All I need  for now...

**Reply by TheEdoRan** [reply]
Implemented `executeOnMount` functionality for hooks in [v7.3.0](https://github.com/TheEdoRan/next-safe-action/releases/tag/v7.3.0). Docs [here](https://next-safe-action.dev/docs/execution/hooks/hook-base-utils#executeonmount).
